It's possible to create aliasing mutable references using safe code by constructing custom queries

**Bevy version**

5bd6deb9749067143efb2dbb681448d94af54e83 (current master)

**Operating system & version**

Windows 10

**What you did**

I've been digging through the ecs code, and the fact that you can construct queries safely means that bevy lets you make and run incompatible queries without unsafe code, which can result in aliasing mutable pointers etc.

The following thread local system, run on a world with a single entity that has a Transform shows the mutable access happening:
```
fn query_test(world: &mut World, resources: &mut Resources) {
    let access = TypeAccess::new(vec![], vec![]);

    let mut query1 = Query::<(Entity, &mut Transform)>::new(world, &access);
    let mut query2 = Query::<(Entity, &mut Transform)>::new(world, &access);

    for (entity1, mut transform1) in query1.iter_mut() {
        for (entity2, mut transform2) in query2.iter_mut() {
            dbg!(transform1.translation);

            transform1.translation += Vec3::new(1.0, 0.0, 0.0);
            dbg!(transform2.translation);
        }
    }
}
```
And the log output:
```
[src\main.rs:45] transform1.translation = Vec3(
    0.0,
    0.0,
    0.0,
)
[src\main.rs:48] transform2.translation = Vec3(
    1.0,
    0.0,
    0.0,
)
```

Would it make sense to hide this method (or make it private) from users too?
So far in thread local systems, I have only accessed queries through `World`, which correctly errors.
```rust
fn system(world: &mut World, _resources: &mut Resources) {
    let mut query1 = world.query_mut::<(Entity, &mut Transform)>();
    let mut query2 = world.query_mut::<(Entity, &mut Transform)>();
    for _ in query1 {}
}
```
```
error[E0499]: cannot borrow `*world` as mutable more than once at a time
  --> src\main.rs:20:22
   |
19 |     let mut query1 = world.query_mut::<(Entity, &mut Transform)>();
   |                      ----- first mutable borrow occurs here
20 |     let mut query2 = world.query_mut::<(Entity, &mut Transform)>();
   |                      ^^^^^ second mutable borrow occurs here
21 |     for _ in query1 {}
   |              ------ first borrow later used here
```

--- a/crates/bevy_ecs/src/system/query/mod.rs
+++ b/crates/bevy_ecs/src/system/query/mod.rs
@@ -31,7 +31,7 @@ impl<'a, Q: HecsQuery> Query<'a, Q> {
     /// This will create a Query that could violate memory safety rules. Make sure that this is only called in
     /// ways that ensure the Queries have unique mutable access.
     #[inline]
-    pub unsafe fn new(
+    pub(crate) unsafe fn new(
         world: &'a World,
         component_access: &'a TypeAccess<ArchetypeComponent>,
     ) -> Self {

--- a/crates/bevy_ecs/src/system/query/mod.rs
+++ b/crates/bevy_ecs/src/system/query/mod.rs
@@ -59,6 +59,15 @@ impl<'a, Q: HecsQuery> Query<'a, Q> {
         unsafe { self.world.query_unchecked() }
     }
 
+    /// Iterates over the query results
+    /// # Safety
+    /// This allows aliased mutability. You must make sure this call does not result in multiple mutable references to the same component
+    #[inline]
+    pub unsafe fn iter_unsafe(&self) -> QueryIter<'_, Q> {
+        // SAFE: system runs without conflicts with other systems. same-system queries have runtime borrow checks when they conflict
+        self.world.query_unchecked()
+    }
+
     #[inline]
     pub fn par_iter(&self, batch_size: usize) -> ParIter<'_, Q>
     where
@@ -97,6 +106,19 @@ impl<'a, Q: HecsQuery> Query<'a, Q> {
                 .query_one_unchecked::<Q>(entity)
                 .map_err(|_err| QueryError::NoSuchEntity)
         }
+    }
+
+    /// Gets the query result for the given `entity`
+    /// # Safety
+    /// This allows aliased mutability. You must make sure this call does not result in multiple mutable references to the same component
+    #[inline]
+    pub unsafe fn get_unsafe(
+        &self,
+        entity: Entity,
+    ) -> Result<<Q::Fetch as Fetch>::Item, QueryError> {
+        self.world
+            .query_one_unchecked::<Q>(entity)
+            .map_err(|_err| QueryError::NoSuchEntity)
     }
 
     /// Gets a reference to the entity's component of the given type. This will fail if the entity does not have
@@ -145,6 +167,19 @@ impl<'a, Q: HecsQuery> Query<'a, Q> {
         } else {
             Err(QueryError::CannotWriteArchetype)
         }
+    }
+
+    /// Gets a mutable reference to the entity's component of the given type. This will fail if the entity does not have
+    /// the given component type
+    /// # Safety
+    /// This allows aliased mutability. You must make sure this call does not result in multiple mutable references to the same component
+    pub unsafe fn get_component_unsafe<T: Component>(
+        &self,
+        entity: Entity,
+    ) -> Result<Mut<'_, T>, QueryError> {
+        self.world
+            .get_mut_unchecked(entity)
+            .map_err(QueryError::ComponentError)
     }
 
     pub fn removed<C: Component>(&self) -> &[Entity] {

--- a/crates/bevy_ecs/src/system/query/mod.rs
+++ b/crates/bevy_ecs/src/system/query/mod.rs
@@ -59,15 +59,6 @@ impl<'a, Q: HecsQuery> Query<'a, Q> {
         unsafe { self.world.query_unchecked() }
     }
 
-    /// Iterates over the query results
-    /// # Safety
-    /// This allows aliased mutability. You must make sure this call does not result in multiple mutable references to the same component
-    #[inline]
-    pub unsafe fn iter_unsafe(&self) -> QueryIter<'_, Q> {
-        // SAFE: system runs without conflicts with other systems. same-system queries have runtime borrow checks when they conflict
-        self.world.query_unchecked()
-    }
-
     #[inline]
     pub fn par_iter(&self, batch_size: usize) -> ParIter<'_, Q>
     where
@@ -106,19 +97,6 @@ impl<'a, Q: HecsQuery> Query<'a, Q> {
                 .query_one_unchecked::<Q>(entity)
                 .map_err(|_err| QueryError::NoSuchEntity)
         }
-    }
-
-    /// Gets the query result for the given `entity`
-    /// # Safety
-    /// This allows aliased mutability. You must make sure this call does not result in multiple mutable references to the same component
-    #[inline]
-    pub unsafe fn get_unsafe(
-        &self,
-        entity: Entity,
-    ) -> Result<<Q::Fetch as Fetch>::Item, QueryError> {
-        self.world
-            .query_one_unchecked::<Q>(entity)
-            .map_err(|_err| QueryError::NoSuchEntity)
     }
 
     /// Gets a reference to the entity's component of the given type. This will fail if the entity does not have
@@ -167,19 +145,6 @@ impl<'a, Q: HecsQuery> Query<'a, Q> {
         } else {
             Err(QueryError::CannotWriteArchetype)
         }
-    }
-
-    /// Gets a mutable reference to the entity's component of the given type. This will fail if the entity does not have
-    /// the given component type
-    /// # Safety
-    /// This allows aliased mutability. You must make sure this call does not result in multiple mutable references to the same component
-    pub unsafe fn get_component_unsafe<T: Component>(
-        &self,
-        entity: Entity,
-    ) -> Result<Mut<'_, T>, QueryError> {
-        self.world
-            .get_mut_unchecked(entity)
-            .map_err(QueryError::ComponentError)
     }
 
     pub fn removed<C: Component>(&self) -> &[Entity] {

--- a/crates/bevy_ecs/src/system/query/query_set.rs
+++ b/crates/bevy_ecs/src/system/query/query_set.rs
@@ -20,7 +20,7 @@ impl<T: QueryTuple> QuerySet<T> {
     /// # Safety
     /// This will create a set of Query types that could violate memory safety rules. Make sure that this is only called in
     /// ways that ensure the Queries have unique mutable access.
-    pub unsafe fn new(world: &World, component_access: &TypeAccess<ArchetypeComponent>) -> Self {
+    pub(crate) unsafe fn new(world: &World, component_access: &TypeAccess<ArchetypeComponent>) -> Self {
         QuerySet {
             value: T::new(world, component_access),
         }

--- a/crates/bevy_ecs/src/system/query/query_set.rs
+++ b/crates/bevy_ecs/src/system/query/query_set.rs
@@ -20,7 +20,10 @@ impl<T: QueryTuple> QuerySet<T> {
     /// # Safety
     /// This will create a set of Query types that could violate memory safety rules. Make sure that this is only called in
     /// ways that ensure the Queries have unique mutable access.
-    pub(crate) unsafe fn new(world: &World, component_access: &TypeAccess<ArchetypeComponent>) -> Self {
+    pub(crate) unsafe fn new(
+        world: &World,
+        component_access: &TypeAccess<ArchetypeComponent>,
+    ) -> Self {
         QuerySet {
             value: T::new(world, component_access),
         }

